src/cargo/ops/cargo_package/mod.rs

```diff
@@ -107,6 +108,28 @@ enum FileContents {
     Generate(GeneratedFile),
 }
 
+impl serde::ser::Serialize for FileContents {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::ser::Serializer,
+    {
+        use serde::ser::SerializeMap as _;
+        match self {
+            FileContents::Copy(path) => {
+                let mut map = serializer.serialize_map(Some(2))?;
+                map.serialize_entry("kind", "copy")?;
+                map.serialize_entry("path", path)?;
+                map.end()
+            }
+            FileContents::Generate(_) => {
+                let mut map = serializer.serialize_map(Some(1))?;
+                map.serialize_entry("kind", "generate")?;
+                map.end()
+            }
+        }
+    }
+}
```

Comment:
Why is this hand written?  If its for `kind`, isn't that just `#[serde(tag = "kind")]`?

Reply:
Fixed, along with the schema extraction!